db/memtable.cc

```diff
@@ -532,38 +532,296 @@ Status MemTable::VerifyEncodedEntry(Slice encoded,
       .GetStatus();
 }
 
-Status MemTable::Add(SequenceNumber s, ValueType type,
-                     const Slice& key, /* user key */
-                     const Slice& value,
-                     const ProtectionInfoKVOS64* kv_prot_info,
-                     bool allow_concurrent,
-                     MemTablePostProcessInfo* post_process_info, void** hint) {
+bool MemTable::InsertKey(std::unique_ptr<MemTableRep>& table, KeyHandle handle,
+                         SequenceNumber s, ValueType type,
+                         const Slice& user_key, /* user key */
+                         const Slice& value, uint32_t encoded_len,
+                         uint32_t* inserted_entries, uint32_t* inserted_len,
+                         std::function<bool(MemTableRep*, KeyHandle)> insert,
+                         bool allow_concurrent) {
+  if (type != kTypeRangeDeletion) {
+    *inserted_entries = 1;
+    *inserted_len = encoded_len;
+    return insert(table.get(), handle);
+  }
+
+  Slice key = user_key;
+
+  if (allow_concurrent) {
+    range_del_table_lock_.lock();
+  }
+  std::unique_ptr<MemTableIterator> iter(
+      new MemTableIterator(*this, ReadOptions(), nullptr /* arena */,
+                           true /* use_range_del_table */));
+  if (iter == nullptr) {
+    if (allow_concurrent) range_del_table_lock_.unlock();
+    return false;
+  }
+
+  // remove the duplicates in to_insert
+  auto cmp = [](const std::pair<Slice, Slice>& a,
+                const std::pair<Slice, Slice>& b) {
+    if (std::get<0>(a).data() < std::get<0>(b).data()) {
+      return true;
+    } else if (std::get<0>(a).data() > std::get<0>(b).data()) {
+      return false;
+    } else if (std::get<1>(a).data() < std::get<1>(b).data()) {
+      return true;
+    } else {
+      return false;
+    }
+  };
+  ParsedInternalKey inkey(key, 0, kTypeRangeDeletion);
+  std::string key_with_seq;
+  AppendInternalKey(&key_with_seq, inkey);
+  iter->SeekForPrev(Slice(key_with_seq.data(), key_with_seq.size()));
+  if (iter->Valid()) {
+    ParsedInternalKey maxkey(ExtractUserKey(iter->key()), kMaxSequenceNumber,
+                             kTypeRangeDeletion);
+    std::string k;
+    AppendInternalKey(&k, maxkey);
+    iter->Seek(Slice(k.data(), k.size()));
+  } else {
+    iter->Seek(Slice(key_with_seq.data(), key_with_seq.size()));
+  }
+
+  auto const& user_comparator = comparator_.comparator.user_comparator();
+  Slice last_key;
+  // start, end, tombstone seqnum
+  std::map<std::pair<Slice, Slice>,
+           std::set<SequenceNumber, std::greater<SequenceNumber>>,
+           decltype(cmp)>
```

Comment:
With the current algorithm for insertion, I don't think we insert any duplicate [start, end) pair? Plus the algorithm can guarantee that sequence numbers in a memtable entry are always sorted. So we can use vector for both map and set here?